subsys/spm/secure_services.c

```diff
@@ -52,6 +52,31 @@ int spm_secure_services_init(void)
 	return err;
 }
 
+#ifdef CONFIG_SPM_SERVICE_READ_FICR
+
+#define FICR_OFFSET_PUBLIC_START        0x204
+#define FICR_OFFSET_PUBLIC_END          0xC1C
+#define FICR_OFFSET_RESTRICTED_START    0x130
+#define FICR_OFFSET_RESTRICTED_END      0x134
+
+__TZ_NONSECURE_ENTRY_FUNC
+int spm_request_read_ficr(u16_t offset, u32_t *destination)
+{
+	if (destination == NULL) {
+		return -EINVAL;
+	}
+
+	if ((offset >= FICR_OFFSET_PUBLIC_START &&
+		offset <= FICR_OFFSET_PUBLIC_END) ||
+		(offset >= FICR_OFFSET_RESTRICTED_START &&
+		offset <= FICR_OFFSET_RESTRICTED_END)) {
+		*destination = *(volatile u32_t *)(0x00FF0000 + offset);
```

Comment:
Can you use some nRF SOC or Dts define for FICR base address

Reply:
Per your questions:
_1. Question: In what way this secure service is different from spm_request_read() which effectively reads some secure memory?_

With spm_request_read() you can only read from the memory starting from PM_MCUBOOT_PAD_ADDRESS to PM_MCUBOOT_PAD_ADDRESS + PM_MCUBOOT_PAD_SIZE. FICR address is outside this range, therefore it is needed. For the reason why FICR need to be accessed from within application, it would be best if we could continue with discussion in private messages.

_2. Can you use some nRF SOC or Dts define for FICR base address?_

In system_nrf9160.c, FICR is accessed by physical address (ref: [here](https://github.com/NordicPlayground/fw-nrfconnect-zephyr/blob/1e34490406b4231fa79558d2759264397fbefa4c/ext/hal/nordic/nrfx/mdk/system_nrf9160.c#L183)). If it is OK there, then it should also be OK here.

Yet, it is a good idea to use macros from nrf9160.h (i.e. NRF_FICR_S_BASE). It is do-able, and will be changed in the next upload.

Reply:
>With spm_request_read() you can only read from the memory starting from PM_MCUBOOT_PAD_ADDRESS to PM_MCUBOOT_PAD_ADDRESS + PM_MCUBOOT_PAD_SIZE. FICR address is outside this range, therefore it is needed

@pkchan why don't you refactor the existing spm_request_read() to work for any address, offset, and size? After all, it is all about reading some secure memory.

Reply:
>In system_nrf9160.c, FICR is accessed by physical address (ref: here). If it is OK there, then it should also be OK here.

system_nrf9160.c is a source file from the nRFx project, that is supposed to work, also, outside the context of Zephyr, even, with bare-metal applications. In Zephyr, however, we adopt a policy for having human-readable macros as much as possible. That's why we have DTS to generate peripheral base addresses (in case we do not take them directly from vendor header, e.g. MDK).